lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -911,30 +942,466 @@ class ConvertAtenNativeLayerNormOp
+
+    if (meanType.getRank() != inputRank || rSTDType.getRank() != inputRank)
+      return rewriter.notifyMatchFailure(
+          op, "input, rstd and mean must be of same rank");
+
+    if (inputRank < normalizedShapeRank)
+      return rewriter.notifyMatchFailure(
+          op, "input cannot have rank less than normalized_shape");
+    if (normalizedShapeRank < 1)
+      return rewriter.notifyMatchFailure(
+          op, "normalized_shape must have rank atleast 1");
+
+    // Check the dimensions of weight and bias to match the dimensions of
+    // normalized_shape.
+    SmallVector<Value> weightAndBiasShape;
+    for (auto en : enumerate((normalizedShapeSizesInt))) {
+      auto index = en.index();
+      auto inputDim = getDimOp(rewriter, loc, input, index + axis);
+      auto weightDim = getDimOp(rewriter, loc, weight, index);
+      auto biasDim = getDimOp(rewriter, loc, bias, index);
+
+      auto expectedSize = en.value();
+      checkDimEqualHelper(rewriter, loc, inputDim, expectedSize);
+      checkDimEqualHelper(rewriter, loc, weightDim, expectedSize);
+      checkDimEqualHelper(rewriter, loc, biasDim, expectedSize);
+      weightAndBiasShape.push_back(weightDim);
+    }
+
+    // Check the dimensions of mean and rSTD to match with the input dims.
+    SmallVector<Value> meanAndSTDShape;
+    for (unsigned i = 0; i < axis; i++) {
+      auto meanDim = getDimOp(rewriter, loc, mean, i);
+      auto rSTDDim = getDimOp(rewriter, loc, rSTD, i);
+      auto inputDim = getDimOp(rewriter, loc, input, i);
+
+      checkDimEqualHelper(rewriter, loc, meanDim, inputDim);
+      checkDimEqualHelper(rewriter, loc, rSTDDim, inputDim);
+      meanAndSTDShape.push_back(inputDim);
+    }
+
+    SmallVector<Value> inputShape = meanAndSTDShape;
+    inputShape.append(weightAndBiasShape);
+
+    // Calculate input_hat = (input - mean) * rSTD
+    AffineMap inputAffineMap = rewriter.getMultiDimIdentityMap(inputRank);
+    SmallVector<AffineExpr> meanAndVarAffineExpr;
+    for (unsigned i = 0; i < axis; i++)
+      meanAndVarAffineExpr.push_back(mlir::getAffineDimExpr(i, context));
+    auto meanAndVarAffineMap = AffineMap::get(
+        /*dimCount=*/inputRank, /*symbolCount=*/0, meanAndVarAffineExpr,
+        context);
+    Value inputHatTensor =
+        rewriter.create<linalg::InitTensorOp>(loc, inputShape, elemTy);
+    SmallVector<StringRef> inputIteratorType(inputRank,
+                                             getParallelIteratorTypeName());
+    SmallVector<StringRef> reshapedMeanVarIteratorType(
+        inputRank, getParallelIteratorTypeName());
+    Value reshapedMeanTensor =
+        rewriter.create<linalg::InitTensorOp>(loc, meanAndSTDShape, elemTy);
+    Value reshapedRSTDTensor =
+        rewriter.create<linalg::InitTensorOp>(loc, meanAndSTDShape, elemTy);
+    SmallVector<AffineMap> meanAndVarIndexingMap{inputAffineMap,
+                                                 meanAndVarAffineMap};
+    Value reshapedMean =
```

Comment:
Why do we need the reshapes for mean and var? The shape is the same before and after the reshape.

Reply:
Actually they are different . We have to reshape the mean to (N,C) from (N,C,1,1,1), to move forward.